### PR TITLE
bpo-43974: Move Py_BUILD_CORE_MODULE into module code (GH-29157)

### PR DESCRIPTION
**Pull request by tiran**
``setup.py`` no longer defines ``Py_BUILD_CORE_MODULE``. Instead every
module defines the macro before ``#include "Python.h"`` unless
``Py_BUILD_CORE_BUILTIN`` is already defined.

``Py_BUILD_CORE_BUILTIN`` is defined for every module that is built by
``Modules/Setup``.

The PR also simplifies ``Modules/Setup``. ``Makefile`` and ``makesetup``
already define ``Py_BUILD_CORE_BUILTIN`` and include ``Modules/internal``
for us.

Signed-off-by: Christian Heimes <christian@python.org>

https://bugs.python.org/issue43974

### CODE REVIEW

File: Modules/_blake2/blake2b_impl.c [hunk 1]
```diff
@@ -13,6 +13,10 @@
  * The blake2s_impl.c is autogenerated from blake2b_impl.c.
  */
 
+#ifndef Py_BUILD_CORE_BUILTIN
+  #define Py_BUILD_CORE_MODULE 1
+#endif
```
Comment (vstinner) [inline Modules/_blake2/blake2b_impl.c]: I have a personal preference to align all `#` on the first column:

```suggestion
#ifndef Py_BUILD_CORE_BUILTIN
#  define Py_BUILD_CORE_MODULE 1
#endif
```

Comment (tiran) [reply Modules/_blake2/blake2b_impl.c]: As you wish!

File: Misc/NEWS.d/next/Build/2021-10-22-14-00-44.bpo-43974.HHZtbx.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+``setup.py`` no longer defines ``Py_BUILD_CORE_MODULE``. Instead every
```
Comment (vstinner) [inline Misc/NEWS.d/next/Build/2021-10-22-14-00-44.bpo-43974.HHZtbx.rst:1]: You may also mention Modules/Setup.

**Review by vstinner — APPROVED** [review]
LGTM. I compared the list of modified files to setup.py. It seems like the PR is exhausitve.

Oh. I didn't know that Modules/makesetup already uses the correct flags for build C extensions depending if it's before or after `*shared*` in Modules/Setup:

```
PY_BUILTIN_MODULE_CFLAGS= $(PY_STDMODULE_CFLAGS) -DPy_BUILD_CORE_BUILTIN
PY_CORE_CFLAGS=	$(PY_STDMODULE_CFLAGS) -DPy_BUILD_CORE
```

File: Misc/NEWS.d/next/Build/2021-10-22-14-00-44.bpo-43974.HHZtbx.rst (hunk 2 above)
Comment (tiran) [reply Misc/NEWS.d/next/Build/2021-10-22-14-00-44.bpo-43974.HHZtbx.rst:1]: > Oh. I didn't know that Modules/makesetup already uses the correct flags for build C extensions depending if it's before or after `*shared*` in Modules/Setup:

There is a problem with ``*shared*``. I created https://bugs.python.org/issue45571 to track the issue.

**Pull request merged**